### PR TITLE
Remove state cwd

### PR DESCRIPTION
**Pull request by gracjan**
Do not suggest or encourage using state in plugins, not in the form of current working directory.

Reference #45.

### CODE REVIEW

File: src/Haskell/Ide/Engine/BasePlugin.hs [hunk 1]
```diff
@@ -67,26 +65,6 @@ baseDescriptor = PluginDescriptor
                         }
           , cmdFunc = commandDetailCmd
           }
-      , Command
-          { cmdDesc = CommandDesc
-                        { cmdName = "pwd"
```
Comment (alanz) [inline src/Haskell/Ide/Engine/BasePlugin.hs]: I think it may be worthwhile keeping this one, but removing "cwd"

**Comment by gracjan** [comment]
@cocreature: Refactored it out to a function.

File: src/Haskell/Ide/Engine/BasePlugin.hs (hunk 1 above)
Comment (gracjan) [inline src/Haskell/Ide/Engine/BasePlugin.hs]: Why? pwd is a constant in this case.

**Comment by alanz** [comment]
There is some logic in setting the initial directory, with a "give up leave it unchanged" option.

I do not believe there is any harm in being able to request the current working directory. It is not a stateful operation.

**Comment by alanz** [comment]
BTW, I am just working through some ghc-mod commands, in particular the "find" command, which works in the context of the current project, which it detects using the find cradle process.

The cradle is defined as the point where the cabal file is found, and the search starts at the current directory.

So in the normal  (non-HIE) case, ghc-mod is started up in the project directory and can then find the cradle by looking in the current directory or moving up.

In the stateless mode, we can work around this by setting the `CtxProject` to require a directory name, and then doing an explicit cd in the ghc-mod plugin, or modifying ghc-mod to take a directory as its root.

We also need to consider the case where there is more than one HIE instance running at a time, attached to different projects.

**Comment by gracjan** [comment]
Yes, `find` command should get an explicit `ctxProject`, an absolute path
to where to begin searching.

2015-11-11 9:14 GMT+01:00 Alan Zimmerman <notifications@github.com>:

**Comment by gracjan** [comment]
And about having a `pwd` without a `cwd`: somebody will come along and contribute the obvious missing part, that is `cwd`. I would not want to give people suggestions for that path.

**Comment by alanz** [comment]
ok, you've worn me down. I'll have to sneak it into a private plugin if I
want to debug.

Maybe we should just log the directory chosen on startup. Then we are both
happy.

**Comment by alanz** [comment]
Can you rebase this, then merge?

**Pull request merged**